### ISSUE TITLE
Fails with UNEXPECTED_MESSAGE or HANDSHAKE_FAILURE where openssl succeeds

### ISSUE THREAD

**Issue by ryanakca** [mirleft/ocaml-tls#413]
I'm unable to connect to certain hosts using ocaml-tls. In particular, connections fail with UNEXPECTED_MESSAGE or HANDSHAKE_FAILURE. I suspect this is due to a bug in ocaml-tls for the following reasons:

1. I am able to connect to these problematic hosts with both s_client and the conduit ocaml library (using OpenSSL as the backend)
2. I am able to connect to other hosts without any issues (so I suspect it isn't user error).

Here's a minimum working example to reproduce the issue, along with the output of s_client.

```
rak@zeta:~/gemini/mwe$ cat dune
(executable
 (public_name mwe)
 (name mwe)
 (libraries base lwt lwt_ppx mirage-crypto-rng.lwt tls.lwt)
 (preprocess (pps lwt_ppx)))
rak@zeta:~/gemini/mwe$ cat mwe.ml
let connect host port =
  let authenticator ~host:_ _ = Ok None in
  let config = Tls.Config.client ~peer_name:host
      ~authenticator ~ciphers:Tls.Config.Ciphers.supported () in
  let%lwt () = Mirage_crypto_rng_lwt.initialize () in
  try%lwt
    let%lwt (ic, oc) = Tls_lwt.connect_ext config (host, port) in
    let%lwt () = Lwt_io.close ic in
    let%lwt () = Lwt_io.close oc in
    let%lwt () = Lwt_io.printl ("Success: " ^ host) in
    Lwt.return ()
  with
  | Tls_lwt.Tls_alert alert ->
    let%lwt () = Lwt_io.printl (host ^ " " ^ Tls.Packet.alert_type_to_string alert) in
    Lwt.return ()

let () =
  Lwt_main.run begin
    let hosts = [ "gemini.circumlunar.space" ; "conman.org" ] in
    let connections = List.map (fun h -> connect h 1965) hosts in
    let%lwt () = Lwt.join connections in
    Lwt.return ()
  end
rak@zeta:~/gemini/mwe$ dune build && ../_build/default/mwe/mwe.exe
Entering directory '/home/rak/gemini'
gemini.circumlunar.space UNEXPECTED_MESSAGE
conman.org HANDSHAKE_FAILURE
rak@zeta:~/gemini/mwe$ echo "" | openssl s_client -servername gemini.circumlunar.space -connect gemini.circumlunar.space:1965
CONNECTED(00000003)
depth=2 O = Digital Signature Trust Co., CN = DST Root CA X3
verify return:1
depth=1 C = US, O = Let's Encrypt, CN = Let's Encrypt Authority X3
verify return:1
depth=0 CN = gemini.circumlunar.space
verify return:1
---
Certificate chain
 0 s:CN = gemini.circumlunar.space
   i:C = US, O = Let's Encrypt, CN = Let's Encrypt Authority X3
 1 s:C = US, O = Let's Encrypt, CN = Let's Encrypt Authority X3
   i:O = Digital Signature Trust Co., CN = DST Root CA X3
---
Server certificate
-----BEGIN CERTIFICATE-----
MIIFZzCCBE+gAwIBAgISA6+yLfYdxyKR9LfDdFVyu39PMA0GCSqGSIb3DQEBCwUA
MEoxCzAJBgNVBAYTAlVTMRYwFAYDVQQKEw1MZXQncyBFbmNyeXB0MSMwIQYDVQQD
ExpMZXQncyBFbmNyeXB0IEF1dGhvcml0eSBYMzAeFw0yMDA1MDMwNjI2MzJaFw0y
MDA4MDEwNjI2MzJaMCMxITAfBgNVBAMTGGdlbWluaS5jaXJjdW1sdW5hci5zcGFj
ZTCCASIwDQYJKoZIhvcNAQEBBQADggEPADCCAQoCggEBAKktKKk0TqnQPpfw8B98
EpehzXSNDlZIIqiCx0f5F2tjOk1bMRajnDP3Ql52ajxmmryszhlvXFEY3URXQtTK
9nlwwJtj/IC8j7qjrpMr0ZYYMzNycbhkWLGBBhpGB0oxqrDLE6A1vgQsQZJAUrMa
XSYnZXKLd3JLYk/yN+7dHiRSBSXqKP9VuSPfCQf8mSTi0SN6j3dbGyAYtXu6z2Uz
uvJ5KYgsmEAd41DqDE85WTUu0Slkbnji4WeixIkp4MbpvPCFciX1fqWzW1Vb1C/+
JXL4OV4rKuJ0M7Ibb6wYqAXlfWdMSu0oTeQ0s1t9CKVt9yqu0RWUJKm4nnliJLYh
gFECAwEAAaOCAmwwggJoMA4GA1UdDwEB/wQEAwIFoDAdBgNVHSUEFjAUBggrBgEF
BQcDAQYIKwYBBQUHAwIwDAYDVR0TAQH/BAIwADAdBgNVHQ4EFgQUJDgI6yAYMRbO
bpmMzJv7klIEscQwHwYDVR0jBBgwFoAUqEpqYwR93brm0Tm3pkVl7/Oo7KEwbwYI
KwYBBQUHAQEEYzBhMC4GCCsGAQUFBzABhiJodHRwOi8vb2NzcC5pbnQteDMubGV0
c2VuY3J5cHQub3JnMC8GCCsGAQUFBzAChiNodHRwOi8vY2VydC5pbnQteDMubGV0
c2VuY3J5cHQub3JnLzAjBgNVHREEHDAaghhnZW1pbmkuY2lyY3VtbHVuYXIuc3Bh
Y2UwTAYDVR0gBEUwQzAIBgZngQwBAgEwNwYLKwYBBAGC3xMBAQEwKDAmBggrBgEF
BQcCARYaaHR0cDovL2Nwcy5sZXRzZW5jcnlwdC5vcmcwggEDBgorBgEEAdZ5AgQC
BIH0BIHxAO8AdgBvU3asMfAxGdiZAKRRFf93FRwR2QLBACkGjbIImjfZEwAAAXHZ
be1OAAAEAwBHMEUCIAL4Mi1nRLZeBZj5Fb4YuLWD0NSMwXmrzqqU492bQbZnAiEA
9NrEo77dXg/3fE/rpyroaMlgLcOcJhvPAdsise9OjFkAdQAHt1wb5X1o//Gwxh0j
Fce65ld8V5S3au68YToaadOiHAAAAXHZbe1RAAAEAwBGMEQCIFHCB2wSWci/Rjnw
TCoLEN7VlCXsCrK6ZuR4LAUy2p7tAiAe759yZTxvDIabNav7bq7V7REKTNSI4ozF
CK6k9uNaeTANBgkqhkiG9w0BAQsFAAOCAQEAHnG7V2rwy3Yhzttn7Iuor7fsVAZ/
3OkI0Y6RJpYQpThUYfyUdGc2H6E1CW8uIihFOqq8/EQ+6sj7PnDnncSO6fy42zwv
alIubUdnRGYA6cNEtx0pad/UPdFBn73luDABKO9nePUAQEL5sIwCrW9SIPOtlKpW
wi2QMnH0Vtb3JlCOc/nPgauadUUIqpGvxyeF1Rd4ZnPWr/zvgKk0JRay2WZ30rlD
KrM8PkHOyNy7Mq74V12EUwC1S2l1QvwCv7xCpBPOnvaD8U7bEZRmoHC3TQ4p4JxH
483ty9vieHJSKTZf2tZytVjUoU3fZCY5tF+und10gmtRdsua15J+qiPh7A==
-----END CERTIFICATE-----
subject=CN = gemini.circumlunar.space

issuer=C = US, O = Let's Encrypt, CN = Let's Encrypt Authority X3

---
No client certificate CA names sent
Client Certificate Types: RSA sign, ECDSA sign
Requested Signature Algorithms: RSA+SHA256:ECDSA+SHA256:RSA+SHA384:ECDSA+SHA384:RSA+SHA512:ECDSA+SHA512:RSA+SHA1:ECDSA+SHA1
Shared Requested Signature Algorithms: RSA+SHA256:ECDSA+SHA256:RSA+SHA384:ECDSA+SHA384:RSA+SHA512:ECDSA+SHA512
Peer signing digest: SHA256
Peer signature type: RSA
Server Temp Key: X25519, 253 bits
---
SSL handshake has read 3169 bytes and written 421 bytes
Verification: OK
---
New, TLSv1.2, Cipher is ECDHE-RSA-AES256-GCM-SHA384
Server public key is 2048 bit
Secure Renegotiation IS supported
Compression: NONE
Expansion: NONE
No ALPN negotiated
SSL-Session:
    Protocol  : TLSv1.2
    Cipher    : ECDHE-RSA-AES256-GCM-SHA384
    Session-ID: 62ADA0D841BBCB2CF3AB90FB75631A74AD4C7D209E522ED300D41C31B5389972
    Session-ID-ctx:
    Master-Key: D3EAEA2D0218C4F3C78B1A5CEE30D59B7A1958A97DF77857192E008D742D2EE8FB3426FC1E789F2227AC1A9D627FE11F
    PSK identity: None
    PSK identity hint: None
    SRP username: None
    TLS session ticket:
    0000 - 1d 93 67 88 6f 18 e5 55-07 00 a4 06 40 62 67 eb   ..g.o..U....@bg.
    0010 - e9 7d 06 ce cf b1 3d 14-0b ce f3 9c 31 1f ea 49   .}....=.....1..I
    0020 - a3 5f a5 6b bf ce 53 71-3d 15 08 1c b1 55 89 2e   ._.k..Sq=....U..
    0030 - 7a 0c 33 3a 20 d7 2b 4d-cf 50 08 a8 07 1f 77 0d   z.3: .+M.P....w.
    0040 - 3d 9e f1 33 1b f0 9a d0-0c 7e 18 b1 d0 e8 97 27   =..3.....~.....'
    0050 - 94 95 c2 38 46 35 a1 b6-e1 78 a4 53 b5 b1 2a ef   ...8F5...x.S..*.
    0060 - ce 44 39 36 90 e4 6f 42-f5 e3 21 26 2b c5 cb da   .D96..oB..!&+...
    0070 - 20 7a a6 d3 d9 ec 1c 5b-                           z.....[

    Start Time: 1591673629
    Timeout   : 7200 (sec)
    Verify return code: 0 (ok)
    Extended master secret: no
---
DONE
rak@zeta:~/gemini/mwe$ echo "" | openssl s_client -servername conman.org -connect conman.org:1965
CONNECTED(00000003)
depth=0 C = US, ST = FL, L = Boca Raton, O = Conman Labs, OU = R&D, CN = gemini.conman.org, emailAddress = sean@conman.org
verify error:num=20:unable to get local issuer certificate
verify return:1
depth=0 C = US, ST = FL, L = Boca Raton, O = Conman Labs, OU = R&D, CN = gemini.conman.org, emailAddress = sean@conman.org
verify error:num=21:unable to verify the first certificate
verify return:1
---
Certificate chain
 0 s:C = US, ST = FL, L = Boca Raton, O = Conman Labs, OU = R&D, CN = gemini.conman.org, emailAddress = sean@conman.org
   i:C = US, ST = FL, O = Conman Laboratories, OU = Security Division, CN = Conman Laboratories CA, emailAddress = ca@conman.org
---
Server certificate
-----BEGIN CERTIFICATE-----
MIIGHDCCBASgAwIBAgICEAkwDQYJKoZIhvcNAQELBQAwgZMxCzAJBgNVBAYTAlVT
MQswCQYDVQQIDAJGTDEcMBoGA1UECgwTQ29ubWFuIExhYm9yYXRvcmllczEaMBgG
A1UECwwRU2VjdXJpdHkgRGl2aXNpb24xHzAdBgNVBAMMFkNvbm1hbiBMYWJvcmF0
b3JpZXMgQ0ExHDAaBgkqhkiG9w0BCQEWDWNhQGNvbm1hbi5vcmcwHhcNMTkwNjIx
MjMyMjE1WhcNMjAwNjIwMjMyMjE1WjCBjzELMAkGA1UEBhMCVVMxCzAJBgNVBAgM
AkZMMRMwEQYDVQQHDApCb2NhIFJhdG9uMRQwEgYDVQQKDAtDb25tYW4gTGFiczEM
MAoGA1UECwwDUiZEMRowGAYDVQQDDBFnZW1pbmkuY29ubWFuLm9yZzEeMBwGCSqG
SIb3DQEJARYPc2VhbkBjb25tYW4ub3JnMIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8A
MIIBCgKCAQEArWGm4XPHQwMtcYDtn1t4DIJzvFvxEG9WJQYXi/qC676NojmY4yGD
bGdkWSbl/xaG2zx+X5vk1YLCtBwQ5R2fBKDXG+PsXFZA/F1vHB9HRejrObPRHgL6
tJwSmAgteZJfzlDqGEUi33NZAD8AItvKz7YhyxhgaMZvBf6wEgIi4qL5DRXg3OaY
+JtTVLKAvIEVbBisrkaF69JL31bbg6ogSaDHCfu3bL5k9h4aBNfDSVbve5XQhsJm
5fpr7+2F3Lny5l5vUxqpnZ27No9m5pI//bIk7pZfwfa3MkQwYxtg2CJ8/5lMtzom
cEQvEG888S2ZJl/9+TfLoKD+BLlDZQ0iYwIDAQABo4IBejCCAXYwCQYDVR0TBAIw
ADARBglghkgBhvhCAQEEBAMCBkAwNAYJYIZIAYb4QgENBCcWJUxpYnJlU1NMIEdl
bmVyYXRlZCBTZXJ2ZXIgQ2VydGlmaWNhdGUwHQYDVR0OBBYEFL6mDBRUDRqfgmzc
hrtvqhgw6nwzMIHbBgNVHSMEgdMwgdCAFLzp1Z8QQY4qzq9NpoAeY6Lkr6UmoYGz
pIGwMIGtMQswCQYDVQQGEwJVUzELMAkGA1UECAwCRkwxEzARBgNVBAcMCkJvY2Eg
UmF0b24xHDAaBgNVBAoME0Nvbm1hbiBMYWJvcmF0b3JpZXMxGjAYBgNVBAsMEVNl
Y3VyaXR5IERpdmlzaW9uMRwwGgYDVQQDDBNDb25tYW4gTGFib3JhdG9yaWVzMSQw
IgYJKoZIhvcNAQkBFhVwb3N0bWFzdGVyQGNvbm1hbi5vcmeCAhAAMA4GA1UdDwEB
/wQEAwIFoDATBgNVHSUEDDAKBggrBgEFBQcDATANBgkqhkiG9w0BAQsFAAOCAgEA
Z97BN1lSylKikbCMH/eUOv3zhPdJkAJgMEWxHXoeYcGnaM20qZFGll0PB9St0Gf9
/ElYxEI65AHb2vlwcgG2OLNUu6Pm7xdNo58JLsVx+HyLCTyNL1GwfHqGFugj9ebj
3oCpDdVN7BkZ7/FpFobfNZNEssYI1iYd4lakFra1JfrrbCzzOHYv3AQR1qoW/493
9VaXNdKb11e9YxkjBlRfcCYOBt1oGrQtQ3xgT66fgCCCjjisvefI2y73dz2lE1iR
xFz0WdBKc244Rk+/5Rk9MICF70AjBkQWu4P3TAS/QwKjYwk9/2k1rHMTbX6us6JD
+HLu7PQ2Y0OHOydsC3bRHq8dBjGYUYM9VbKIf3WjjN9GQB+ethsSQRvdFeM1jNSd
f1OH9LYQOyk4/+r16p9pKaJiniFlcEubnNIhrZQNXAldl7s1uXV733zgV9sLZ2sR
mZUy/AkH1myw2rXE7m7SxggZD1DsO4F91xhGm1jl3LZdWhfl/wKVLUt9cIPgixgV
AwFsRmxhXlUN+yvhfQFT7c7inuczM7h4ImLkfBF0eu4FdPvw6KfhTdns6WiEgiAd
8aZKx/V4qvpzkngRi6OUAyUA1Bzvy6Z2miibYoq66DHxFRyER5/zcsqDx13v6ee/
jSHPzpej+9P+4zhUiSacYrRKhXR3/CBjc5zINPbLNJs=
-----END CERTIFICATE-----
subject=C = US, ST = FL, L = Boca Raton, O = Conman Labs, OU = R&D, CN = gemini.conman.org, emailAddress = sean@conman.org

issuer=C = US, ST = FL, O = Conman Laboratories, OU = Security Division, CN = Conman Laboratories CA, emailAddress = ca@conman.org

---
No client certificate CA names sent
Client Certificate Types: RSA sign, ECDSA sign
Requested Signature Algorithms: RSA-PSS+SHA512:RSA+SHA512:ECDSA+SHA512:RSA-PSS+SHA384:RSA+SHA384:ECDSA+SHA384:RSA-PSS+SHA256:RSA+SHA256:ECDSA+SHA256:RSA+SHA1:ECDSA+SHA1
Shared Requested Signature Algorithms: RSA-PSS+SHA512:RSA+SHA512:ECDSA+SHA512:RSA-PSS+SHA384:RSA+SHA384:ECDSA+SHA384:RSA-PSS+SHA256:RSA+SHA256:ECDSA+SHA256
Peer signing digest: SHA256
Peer signature type: RSA-PSS
Server Temp Key: X25519, 253 bits
---
SSL handshake has read 2042 bytes and written 399 bytes
Verification error: unable to verify the first certificate
---
New, TLSv1.2, Cipher is ECDHE-RSA-CHACHA20-POLY1305
Server public key is 2048 bit
Secure Renegotiation IS supported
Compression: NONE
Expansion: NONE
No ALPN negotiated
SSL-Session:
    Protocol  : TLSv1.2
    Cipher    : ECDHE-RSA-CHACHA20-POLY1305
    Session-ID:
    Session-ID-ctx:
    Master-Key: AD0C0783EDB859FFA64C0FD57A3E7DAD142A4696684729E9FCBF3C477B79AC3FA424D2C9C81D412E3722F6266FA46B49
    PSK identity: None
    PSK identity hint: None
    SRP username: None
    Start Time: 1591673632
    Timeout   : 7200 (sec)
    Verify return code: 21 (unable to verify the first certificate)
    Extended master secret: no
---
DONE
```

**Comment by hannesm** [comment]
In #414 I developed support for ECDHE and TLS < 1.3. If you could try that out, that'd be great (use your code from above, and that branch of tls - `opam pin add tls git+https://github.com/hannesm/ocaml-tls.git#ecdhe-1-2` followed by a recompile (dune clean && dune build) of your code).

**Comment by ryanakca** [comment]
Thanks for the quick response, Hannes!

It looks like this partially fixes the issue. I still get UNEXPECTED_MESSAGE for gemini.circumlunar.space, but conman.org has gone from HANDSHAKE_FAILURE to DECODE_ERROR.

```
rak@zeta:~/gemini/mwe$ opam pin add tls git+https://github.com/hannesm/ocaml-tls.git#ecdhe-1-2
[tls.0.12.0] synchronised from git+https://github.com/hannesm/ocaml-tls.git#ecdhe-1-2
tls is now pinned to git+https://github.com/hannesm/ocaml-tls.git#ecdhe-1-2 (version 0.12.0)

The following actions will be performed:
  ↻ recompile tls              0.12.0*
  ↻ recompile conduit-lwt-unix 2.1.0   [uses tls]
===== ↻ 2 =====
Do you want to continue? [Y/n] y

<><> Gathering sources ><><><><><><><><><><><><><><><><><><><><><><><><><><><><>
[conduit-lwt-unix.2.1.0] found in cache
[tls.0.12.0] synchronised from git+https://github.com/hannesm/ocaml-tls.git#ecdhe-1-2

<><> Processing actions <><><><><><><><><><><><><><><><><><><><><><><><><><><><>
⊘ removed   conduit-lwt-unix.2.1.0
⊘ removed   tls.0.12.0
∗ installed tls.0.12.0
∗ installed conduit-lwt-unix.2.1.0
Done.
rak@zeta:~/gemini/mwe$ dune clean && dune build && ../_build/default/mwe/mwe.exe
Entering directory '/home/rak/gemini'
gemini.circumlunar.space UNEXPECTED_MESSAGE
conman.org DECODE_ERROR
rak@zeta:~/gemini/mwe$
```

**Comment by hannesm** [comment]
thanks for checking, I tested your endpoints manually (with `dune build @all` and `lwt/examples/echo_client.exe`) -- and pushed [a fix](https://github.com/mirleft/ocaml-tls/pull/414/commits/45445169810e870dd0bb9ba54169b98f281acc6c) to the PR.

could you try this out with `opam reinstall tls` (now that you have it pinned to my branch), and report whether this works for you?

**Comment by hannesm** [comment]
With another change for release preparations, you'll need to adapt your code above by inserting a call to `Mirage_crypto_rng_lwt.initialize () >>= fun () ->` in your `let () = ...` inside the `Lwt_main.run`.

**Comment by hannesm** [comment]
the merge of #414 closed this issue, please reopen if you encounter any further issues. the fix has been released as part of 0.12.1 (currently pending at opam-repository https://github.com/ocaml/opam-repository/pull/16625). thanks again for your report.
